Rubberduck.Parsing/VBA/RubberduckParserReimpl.cs

```diff
@@ -0,0 +1,356 @@
+            }
+        }
+
+        private void ReparseRequested(object sender, EventArgs e)
+        {
+            var args = e as ParseRequestEventArgs;
+            if (args.IsFullReparseRequest)
+            {
+                Cancel();
+                ParseAll();
+            }
+            else
+            {
+                Cancel(args.Component);
+                ParseAsync(args.Component, CancellationToken.None, _state.GetRewriter(args.Component));
+            }
+        }
+
+        /// <summary>
+        /// Starts parsing all components of all unprotected VBProjects associated with the VBE-Instance passed to the constructor of this parser instance.
+        /// </summary>
+        private void ParseAll()
+        {
+            var projects = _vbe.VBProjects
+                .Cast<VBProject>()
+                .Where(project => project.Protection == vbext_ProjectProtection.vbext_pp_none);
+
+            var components = projects.SelectMany(p => p.VBComponents.Cast<VBComponent>());
+            // invalidation cleanup should go into ParseAsync?
+            foreach (var invalidated in _componentAttributes.Keys.Except(components))
+            {
+                _componentAttributes.Remove(invalidated);
+            }
+
+            foreach (var vbComponent in components)
+            {
+                while (!_state.ClearDeclarations(vbComponent)) { }
+
+                ParseAsync(vbComponent, CancellationToken.None, _state.GetRewriter(vbComponent));
+            }
+        }
+
+        public Task ParseAsync(VBComponent component, CancellationToken token, TokenStreamRewriter rewriter = null)
+        {
+            // FIXME remove invalidated "things"from _state
+            // this includes: Declarations, Comments, Attributes, "InspectionResults" (ObsoleteCall, ObsoleteLet, EmptyStringLiteral, ArgLists with OneByRef) and possibly more...
+
+            var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(_central.Token, token);
+
+            var task = new Task(() => ParseAsyncInternal(component, linkedTokenSource.Token, rewriter));
+            _currentTasks.Add(component, Tuple.Create(task, linkedTokenSource));
+
+            task.ContinueWith(t => _currentTasks.Remove(component)); // default also executes on cancel
+
+            _state.SetModuleState(component, ParserState.Parsing);
+            task.Start();
+
+            return task;
+        }
+
+        public void Cancel(VBComponent component = null)
+        {
+            // FIXME cancel resolver task if running
+            if (component == null)
+            {
+                _central.Cancel(false);
+            }
+            else
+            {
+                Tuple<Task, CancellationTokenSource> result;
+                if (_currentTasks.TryGetValue(component, out result))
+                {
+                    result.Item2.Cancel(false);
+                    result.Item1.Wait(); // should we do this??
```

Comment:
Waiting for the task will synchronously block until it reaches an end-state -- that seems the opposite of what you intend to do here to me. Assuming the `CancellationTokenSource` is already connected to the `Task` I think you can just omit this line.